src/VisualStudio/Core/Def/Implementation/ProjectTelemetry/VisualStudioProjectTelemetryService.cs

```diff
@@ -0,0 +1,172 @@
+            // Have to catch all exceptions coming through here as this is called from a
+            // fire-and-forget method and we want to make sure nothing leaks out.
+            try
+            {
+                await StartWorkerAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // Cancellation is normal (during VS closing).  Just ignore.
+            }
+            catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+            {
+                // Otherwise report a watson for any other exception.  Don't bring down VS.  This is
+                // a BG service we don't want impacting the user experience.
+            }
+        }
+
+        private async Task StartWorkerAsync(CancellationToken cancellationToken)
+        {
+            _workQueue = new AsyncBatchingWorkQueue<ProjectTelemetryInfo>(
+                TimeSpan.FromSeconds(1),
+                NotifyTelemetryServiceAsync,
+                cancellationToken);
+
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+                return;
+
+            // Pass ourselves in as the callback target for the OOP service.  As it discovers
+            // designer attributes it will call back into us to notify VS about it.
+            _keepAliveSession = await client.TryCreateKeepAliveSessionAsync(
+                WellKnownServiceHubServices.RemoteProjectTelemetryService,
+                callbackTarget: this, cancellationToken).ConfigureAwait(false);
+            if (_keepAliveSession == null)
+                return;
+
+            // Now kick off scanning in the OOP process.
+            var success = await _keepAliveSession.TryInvokeAsync(
+                nameof(IRemoteProjectTelemetryService.ComputeProjectTelemetryAsync),
+                solution: null,
+                arguments: Array.Empty<object>(),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Callback from the OOP service back into us.
+        /// </summary>
+        public Task RegisterProjectTelemetryInfoAsync(ProjectTelemetryInfo info, CancellationToken cancellationToken)
+        {
+            _workQueue!.AddWork(info);
+            return Task.CompletedTask;
+        }
+
+        private async Task NotifyTelemetryServiceAsync(
+            ImmutableArray<ProjectTelemetryInfo> infos, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var _1 = ArrayBuilder<ProjectTelemetryInfo>.GetInstance(out var filteredInfos);
+            using var _2 = PooledHashSet<ProjectId>.GetInstance(out var seenProjectIds);
+
+            for (var i = infos.Length - 1; i >= 0; i--)
+            {
+                var info = infos[i];
+                if (seenProjectIds.Add(info.ProjectId))
```

Comment:
This is going through the infos backwards because the latest values will have the most up to date project information, correct?  And then we don't care what the previously reported projectinfo was and don't need to report it?  This could definitely use a comment.

Reply:
i could have sworn i wrote that comment :)  must have lost it in a refactoring :)